File: uri/1029/1029.cpp
#include <cstdio>

long int f[39];
long int r[39];

long int fib(long int n) {
    if (f[n] != -1)
        return f[n];
    if (n <= 1) {
        f[n] = n;
    } else {
        f[n] = fib(n - 1) + fib(n - 2);
        if (n >= 3)
            r[n] = r[n - 1] + r[n - 2] + 2;
    }
    return f[n];
}

int main() {
    int i, j;
    long int n;
    for (j = 0; j <= 39; j++) {
        f[j] = -1;
        r[j] = -1;
    }

    r[0] = 0;
    r[1] = 0;
    r[2] = 2;

    scanf("%d", &i);
    while (i--) {
        scanf("%ld", &n);
        printf("fib(%ld) = %ld calls = %ld\n", n, r[n], fib(n));
    }
    return 0;
}

File: solutions/beecrowd/1029/1029.clj
(ns main)

(def fib
  (memoize (fn [number]
             (cond
               (<= number 1)
               number
               :else
               (+ (fib (- number 1))
                  (fib (- number 2)))))))

(def fib-calls
  (memoize (fn [number]
             (cond
               (<= number 1)
               0
               :else
               (+ (fib-calls (- number 1))
                  (fib-calls (- number 2))
                  2)))))

(defn print-result [number]
  (let [value (fib number)
        calls (fib-calls number)]
   (printf "fib(%d) = %d calls = %d%n" number calls value)))

(defn main []
  (loop [n (Integer/parseInt (read-line))]
    (print-result (Integer/parseInt (read-line)))
    (when (> n 1)
      (recur (dec n)))))

(main)
